### PR TITLE
HIVE-25522 NullPointerException in TxnHandler

### PR DESCRIPTION
**Pull request by szehon-ho**
### What changes were proposed in this pull request?
* This fixes https://issues.apache.org/jira/browse/HIVE-25522
* There are two options, either make the initialization static and kill HMS if there is an error, or keep it lazy.  Went with second approach as there seem to be db connections that are taken and don't need to be if nobody uses any txnHandler methods.
* Make the initailization setConf method be idempotent by checking each of the static variables it sets and able to resume if a particular variable is not set.
* Also some refactor to push down verbose catch blocks as much as possible. 

### Why are the changes needed?
* See https://issues.apache.org/jira/browse/HIVE-25522

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Running unit tests

### CODE REVIEW

**Comment by pvary** [comment]
Thanks for the patch @szehon-ho!

Did not have time to check the code, but I can imagine us setting the `connPool` to `null` just to signal that we have to reinitialize the whole `TxnHandler` object. If this is the case we might end up with an invalid object where some values are for the old conf, some values are for the new.

I will check the code if I have some time later.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 1]
```diff
@@ -369,32 +369,36 @@ public void setConf(Configuration conf){
     this.conf = conf;
 
     synchronized (TxnHandler.class) {
+      int maxPoolSize = MetastoreConf.getIntVar(conf, ConfVars.CONNECTION_POOLING_MAX_CONNECTIONS);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:372]: minor: maxPoolSize & connectionTimeoutMs could me moved outside of synchronized block

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 2]
```diff
@@ -5567,10 +5570,16 @@ private void removeTxnsFromMinHistoryLevel(Connection dbConn, List<Long> txnids)
     }
   }
 
-  private static synchronized DataSource setupJdbcConnectionPool(Configuration conf, int maxPoolSize, long getConnectionTimeoutMs) throws SQLException {
+  private static synchronized DataSource setupJdbcConnectionPool(Configuration conf, int maxPoolSize, long getConnectionTimeoutMs) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5573]: I don't think it needs to be static and synchronized. The only place where it's used is setConf method under the synchronized block

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 3]
```diff
@@ -5567,10 +5570,16 @@ private void removeTxnsFromMinHistoryLevel(Connection dbConn, List<Long> txnids)
...
     DataSourceProvider dsp = DataSourceProviderFactory.tryGetDataSourceProviderOrNull(conf);
     if (dsp != null) {
-      return dsp.create(conf);
+      try {
+        return dsp.create(conf);
+      } catch (SQLException e) {
+        String msg = "Unable to instantiate JDBC connection pooling, " + e.getMessage();
+        LOG.error(msg);
+        throw new RuntimeException(e);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5581]: Why is this change required? There is handling of checked SQLException in setConf.

Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5581]: Could we please add msg to the exception, i.e. ````throw new RuntimeException(msg, e)````

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 4]
```diff
@@ -369,32 +369,36 @@ public void setConf(Configuration conf){
...
-          dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
+        connPoolMutex = setupJdbcConnectionPool(conf, maxPoolSize + MUTEX_KEY.values().length, getConnectionTimeoutMs);
+      }
+
+      if (dbProduct == null) {
+        try (Connection dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED)) {
           determineDatabaseProduct(dbConn);
-          sqlGenerator = new SQLGenerator(dbProduct, conf);
         } catch (SQLException e) {
           String msg = "Unable to instantiate JDBC connection pooling, " + e.getMessage();
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:393]: Should we update the exception text here as you are handling pooling exceptions inside of setupJdbcConnectionPool?

Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 3 above)
Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5581]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 2 above)
Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 1 above)
Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Done

**Comment by szehon-ho** [comment]
Thanks for the review.  Yea that's another option (or an 'initialized' boolean set only at the end), but I was afraid it would leave behind some connection or things that need cleanup if we partially initialize and re-initialize, though maybe it would not.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 5]
```diff
@@ -368,33 +368,37 @@ public TxnHandler() {
...
     synchronized (TxnHandler.class) {
       if (connPool == null) {
-        Connection dbConn = null;
-        // Set up the JDBC connection pool
-        try {
-          int maxPoolSize = MetastoreConf.getIntVar(conf, ConfVars.CONNECTION_POOLING_MAX_CONNECTIONS);
-          long getConnectionTimeoutMs = 30000;
-          connPool = setupJdbcConnectionPool(conf, maxPoolSize, getConnectionTimeoutMs);
-          /*the mutex pools should ideally be somewhat larger since some operations require 1
+        connPool = setupJdbcConnectionPool(conf, maxPoolSize, getConnectionTimeoutMs);
```
Comment (sunchao) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:375]: I wonder if we should initialize these static fields separately (perhaps in HiveMetaStore) instead of within each handler thread. Otherwise, all the handler threads will try this code and fail inevitably.

Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:375]: Do you mean eager initialization on HMS startup?  I had thought it too and initially did not want to change the behavior and add more connections for potentially nothing for HMS user does not use any Txn functions.  Is it what you mean?  (Not sure all the threads will inevitably fail though, after the first succeeds they will skip)

Comment (sunchao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:375]: > potentially add more connections for for HMS user who do not use any Txn functions

Yea that's true .. I wonder if there is a way to know that the txn feature will be used beforehand.

> Not sure all the threads will inevitably fail though, after the first succeeds they will skip

What I mean is if there is some fatal issue in the DB connection for instance, all the threads will try the same path and fail. It's better to just fail once instead.

Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:375]: Yea it is maybe better..  Though I don't see anything about turning off txn handler, the only thing is :  metastore.txn.store.impl but it always has a default value.

@pvary  @deniskuzZ any thoughts if we prefer eager initialization?

Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:375]: > What I mean is if there is some fatal issue in the DB connection for instance, all the threads will try the same path and fail. It's better to just fail once instead.

By the way in our case it would have recovered as db connection becomes available after awhile :) , but yea hard to generalize the case

**Comment by szehon-ho** [comment]
I tried Chao's suggestion to move the initialization of TxnHandler to be eager.  It is true that going forward more users of Hive metastore are using it's transactional functionalities.

Just fixed the tests, @sunchao @pvary @deniskuzZ if you want to take a look at this approach.  

Or let me know if we want to go back to the lazy init approach.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java [hunk 6]
```diff
@@ -433,6 +433,7 @@ public void init() throws MetaException {
         addAdminUsers();
         currentUrl = MetaStoreInit.getConnectionURL(conf);
       }
+      TxnUtils.initializeTxnStore(conf); // Must be before Metric services start
```
Comment (sunchao) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:436]: hmm will each `HMSHandler` be initialized in its own thread? curious how is this different from the existing code.

Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:436]: Yea so if you trace HiveMetastore.startMetaStore() => RetryingHMSHandler.getProxy => HMSHandler.init() (so this makes it eager initialization) 

I thought putting it here fixes the test, but actually tests are still running so maybe there are other test errors after.

Comment (sunchao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:436]: Oh I see. Makes sense.

Strange that I don't see the latest test results.

**Comment by szehon-ho** [comment]
@kgyrtkirk do you know how to retrigger this build?  (not sure if it timed out or..)

**Comment by sunchao** [comment]
@szehon-ho you can restart it manually - I just did it

**Comment by sunchao** [comment]
@szehon seems test failures are related:
```
Caused by: org.apache.hive.com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: FATAL: remaining connection slots are reserved for non-replication superuser connections
	at org.apache.hive.com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:544) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at org.apache.hive.com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:536) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at org.apache.hive.com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:112) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at org.apache.hive.com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:72) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at org.apache.hadoop.hive.metastore.datasource.HikariCPDataSourceProvider.create(HikariCPDataSourceProvider.java:90) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at org.apache.hadoop.hive.metastore.txn.TxnHandler.setupJdbcConnectionPool(TxnHandler.java:5571) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
	at org.apache.hadoop.hive.metastore.txn.TxnHandler.init(TxnHandler.java:370) ~[hive-exec-4.0.0-SNAPSHOT.jar:4.0.0-SNAPSHOT]
```

**Comment by szehon-ho** [comment]
Thanks, yea somehow I didn't see the test results, but see it now.  Taking a look, looks like there's going to be some impacts to many tests.  This one in particular (TezTPCDS30TBPerfCliDriver) starts a postgres with very limited pool, have to see how to fix it.

**Comment by szehon-ho** [comment]
Hi guys sorry for delay, I took a little look at it.  Seems like test fail because in test environment TxnHandler is initialized more than once but not tearing down when shutdown (seems JVM is recycled).

Actually found something else, in current version of Hive it is now on critical path of startup, since AcidMetricService is added to list of services in https://issues.apache.org/jira/browse/HIVE-24824.  

So seems it's a Hive 3 only bug.

**Comment by szehon-ho** [comment]
Anyway i made a fix for the test, and try again.  I guess it might still be an issue if the user removes AcidMetricService from the configured services.

**Comment by szehon-ho** [comment]
Few tests left to fix, TestTxnXXX that now need to initialize TxnStore manually (init is longer lazy part of getTxnStore).

**Comment by szehon-ho** [comment]
Looks like random failure, as previous run succeeded (intermediate change is only to add a license header), retriggering

**Comment by szehon-ho** [comment]
Test failure doesn't look related.  @sunchao can you take another look when you have time?  Thanks

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 7]
```diff
@@ -5567,10 +5570,16 @@ private void removeTxnsFromMinHistoryLevel(Connection dbConn, List<Long> txnids)
     }
   }
 
-  private static synchronized DataSource setupJdbcConnectionPool(Configuration conf, int maxPoolSize, long getConnectionTimeoutMs) throws SQLException {
+  private static DataSource setupJdbcConnectionPool(Configuration conf, int maxPoolSize, long getConnectionTimeoutMs) {
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Restored

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 8]
```diff
@@ -5567,10 +5570,16 @@ private void removeTxnsFromMinHistoryLevel(Connection dbConn, List<Long> txnids)
...
     DataSourceProvider dsp = DataSourceProviderFactory.tryGetDataSourceProviderOrNull(conf);
     if (dsp != null) {
-      return dsp.create(conf);
+      try {
+        return dsp.create(conf);
+      } catch (SQLException e) {
+        String msg = "Unable to instantiate JDBC connection pooling, " + e.getMessage();
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: Changed (i had just copied the original one)

**Review by sunchao — APPROVED** [review]
LGTM, sorry for the late review @szehon-ho
